[RA2 Ch3] Containers Bare Metal and in VMs (#2417)
cntt-n/CNTT: doc/ref_arch/kubernetes/chapters/chapter03.md

<p align="center"><b>Table 3-1:</b> Categories of applications, requirements for scheduling pods and Kubernetes features</p>

> Kubernetes clusters with above enhancements can run on bare metal servers (directly on Linux without host Operating System) or in virtual machines (VMs, on hypervisor).

Reviewer: "directly on Linux without host Operating System" - do you mean "with" instead of "without"?

Author: > "directly on Linux without host Operating System" - do you mean "with" instead of "without"?

@tomkivlin : Yes. Hope it is more precise as `can run "bare metal" (directly on Linux host Operating System)`. Thanks

> When running in VMs, the following requirements and configuration are needed for performance optimization: CPU Manager for vCPUs in VM. Huge pages enabled in hypervisor and mapped into VM. Hardware Topology Management with hypervisors enabled for NUMA. Node Feature Discovery and Device Plugin Framework for CPU instructions and devices exposed through hypervisor.

Reviewer 2: These seem IaaS layer requirements for the CaaS-on-VM scenario? should we convert this to table format and add to chapter 4?

Author: > These seem IaaS layer requirements for the CaaS-on-VM scenario?
> Can we reword this in plain english and also write the details as chapter 4 specs in addition please?

@rgstori : Yes. Please see if commit [bfef4b9](https://github.com/cntt-n/CNTT/commit/bfef4b99d80468f4662913c4356fa53889f5af14) reads better. Then I'll create another PR for Ch4.

<p align="center"><b>Table 3-1:</b> Categories of applications, requirements for scheduling pods and Kubernetes features</p>

> Kubernetes clusters with above enhancements can run "bare metal" (directly on Linux host Operating System) or in virtual machines (VMs, on hypervisor).

Reviewer 3: @petorre Agree that "without" is wrong but I think we can let the "on" and "servers" to remain

Reviewer 4: suggest rewording 
(directly on Linux host Operating System)
with
(on container run-time environment (CRE))

How much of Linux or other OS in CRE is left to distros.

Author: @pgoyal01 and @arkadykanevsky : Thanks. Please see commit [6f84449](https://github.com/cntt-n/CNTT/pull/2417/commits/6f8444950d4a78c34ac3269c0967f3dcc2a42eb9) where I try to make it more precise.

Reviewer 3: @petorre Thanks. Approved.

<p align="center"><b>Table 3-1:</b> Categories of applications, requirements for scheduling pods and Kubernetes features</p>

Kubernetes clusters using above enhancements can implement worker nodes with "bare metal" servers (running Container Runtime in Linux host Operating System) or with virtual machines (VMs, on hypervisor).

> When running in VMs, the following list of configurations shows what is needed for non-resilient applications:

Reviewer 3: Why "non-resilient"? The elements in the bullets are related to performance.

Author: > Why "non-resilient"? The elements in the bullets are related to performance.

@pgoyal01 : That section is already called "Scheduling Pods with Non-resilient Applications" and starts with "Non-resilient applications are sensitive to platform impairments..." which is more precise description of target apps. With more application resilience, we could get to good levels of "performance" without some of those enhancements.

> * CPU Manager managing vCPUs that hypervisor provides into VM.

Reviewer 3: Typo
```suggestion
* CPU Manager managing vCPUs that hypervisor provides to VMs.
```

> * Huge pages enabled in hypervisor, mapped into VM, enabled in guest OS, and mapped into pod.

Reviewer 3: ```suggestion
* Huge pages enabled in hypervisor, mapped to VM, enabled in guest OS, and mapped to pod.
```

> * Hardware Topology Management with NUMA enabled in hypervisor, mapped into VM, if needed enabled in guest OS, and mapped into pod.

Reviewer 3: NUMA has to be enabled in the BIOS (or hardware) also.

Author: > NUMA has to be enabled in the BIOS (or hardware) also.

@pgoyal01 : Yes, also SMT, and both also for bare metal. I'm not sure that fits in this section. The idea of this was to compare bare metal and VMs.

> * Node Feature Discovery and Device Plugin Framework for CPU instructions and devices enabled in hypervisor, mapped into VM, enabled in guest OS, and mapped into pod.

Reviewer 3: Aren't Node Feature Discovery or Device Plugins independent of the Guest OS? So, they can't be enabled in the Guest OS. Correct?

Author: > Aren't Node Feature Discovery or Device Plugins independent of the Guest OS? So, they can't be enabled in the Guest OS. Correct?

@pgoyal01 : After studying more documentation and thinking about recent projects, with commit [25ac683](https://github.com/cntt-n/CNTT/pull/2417/commits/25ac683a4210b1a8cf86f892cd00458d4f6f2b9e) I added "if needed". For NFD it can depend on kernel module and kernel configuration. For Device Plugin it can require Linux kernel driver. CPU instructions and host-device CNI will work.
